parser/src/ast/mod.rs

```diff
@@ -288,6 +290,90 @@ impl PartialEq for Program {
             && self.integrity_constraints == other.integrity_constraints
     }
 }
+impl fmt::Display for Program {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        writeln!(f, "def {}\n", self.name)?;
+
+        writeln!(f, "trace_columns:")?;
+        for segment in self.trace_columns.iter() {
+            writeln!(f, "    {}", segment)?;
+        }
+        f.write_str("\n")?;
+
+        writeln!(f, "public_inputs:")?;
+        for public_input in self.public_inputs.values() {
+            writeln!(f, "    {}: [{}]", public_input.name, public_input.size)?;
+        }
+        f.write_str("\n")?;
+
+        if let Some(rv) = self.random_values.as_ref() {
+            writeln!(f, "random_values:")?;
+            writeln!(f, "    {}", rv)?;
+            f.write_str("\n")?;
+        }
+
+        if !self.periodic_columns.is_empty() {
+            writeln!(f, "periodic_columns:")?;
+            for (qid, column) in self.periodic_columns.iter() {
+                if qid.module == self.name {
+                    writeln!(
+                        f,
+                        "    {}: {}",
+                        &qid.item,
+                        DisplayList(column.values.as_slice())
+                    )?;
+                } else {
+                    writeln!(f, "    {}: {}", qid, DisplayList(column.values.as_slice()))?;
+                }
+            }
+            f.write_str("\n")?;
+        }
+
+        if !self.constants.is_empty() {
+            for (qid, constant) in self.constants.iter() {
+                if qid.module == self.name {
+                    writeln!(f, "    const {} = {}", &qid.item, &constant.value)?;
+                } else {
+                    writeln!(f, "    const {} = {}", qid, &constant.value)?;
```

Comment:
We don't indent constant declarations in the language, so we probably shouldn't do it here either

Reply:
Ah yep, not sure why I indented these